provider_azure.go

```diff
@@ -0,0 +1,185 @@
+		if err != nil {
+			return nil, fmt.Errorf("unable to get claim sources: %s", err)
+		}
+
+		// Get provider because we'll need to get a new token for microsoft's
+		// graph API, specifically the old graph API
+		provider, err := b.getProvider(b.cachedConfig)
+		if err != nil {
+			return nil, fmt.Errorf("unable to get provider: %s", err)
+		}
+		a.provider = provider
+
+		a.ctx, err = b.createCAContext(b.providerCtx, b.cachedConfig.OIDCDiscoveryCAPEM)
+		if err != nil {
+			return nil, fmt.Errorf("unable to create CA Context: %s", err)
+		}
+
+		azureGroups, err := a.getAzureGroups(b.Logger(), azureClaimSourcesURL, b.cachedConfig)
+		if err != nil {
+			return nil, fmt.Errorf("%q claim not found in token: %v", role.GroupsClaim, err)
+		}
+		groupsClaimRaw = azureGroups
+	}
+	b.Logger().Debug(fmt.Sprintf("groups claim raw is %v", groupsClaimRaw))
+	return groupsClaimRaw, nil
+}
+
+// In Azure, if you are indirectly member of more than 200 groups, they will
+// send _claim_names and _claim_sources instead of the groups, per OIDC Core
+// 1.0, section 5.6.2:
+// https://openid.net/specs/openid-connect-core-1_0.html#AggregatedDistributedClaims
+// In the future this could be used with other providers as well. Example:
+//
+// {
+// 	 "_claim_names": {
+// 	   "groups": "src1"
+// 	 },
+// 	 "_claim_sources": {
+// 	   "src1": {
+// 	     "endpoint": "https://graph.windows.net...."
+// 	   }
+//   }
+// }
+//
+// For this to work, "profile" should be set in "oidc_scopes" in the vault oidc role.
+//
+func (a *AzureProvider) getClaimSource(logger log.Logger, allClaims map[string]interface{}, role *jwtRole) (string, error) {
+	// Get the source key for the groups claim
+	name := fmt.Sprintf("/%s/%s", claimNamesField, role.GroupsClaim)
+	groupsClaimSource := getClaim(logger, allClaims, name)
+	if groupsClaimSource == nil {
+		return "", fmt.Errorf("unable to locate groups claim %q in %s", role.GroupsClaim, claimNamesField)
+	}
+	// Get the endpoint source for the groups claim
+	endpoint := fmt.Sprintf("/%s/%s/endpoint", claimSourcesField, groupsClaimSource.(string))
+	val := getClaim(logger, allClaims, endpoint)
+	if val == nil {
+		return "", fmt.Errorf("unable to locate %s in claims", endpoint)
+	}
+	logger.Debug(fmt.Sprintf("found Azure Graph API endpoint for group membership: %v", val))
+	return fmt.Sprintf("%v", val), nil
+}
+
+// Fetch user groups from the Azure AD Graph API
+func (a *AzureProvider) getAzureGroups(logger log.Logger, groupsURL string, c *jwtConfig) (interface{}, error) {
+	urlParsed, err := url.Parse(groupsURL)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse distributed groups source url %s: %s", groupsURL, err)
+	}
+	token, err := a.getAzureToken(logger, c, urlParsed.Host)
+	if err != nil {
+		return nil, fmt.Errorf("unable to get token: %s", err)
+	}
+	payload := strings.NewReader("{\"securityEnabledOnly\": false}")
+	req, _ := http.NewRequest("POST", groupsURL, payload)
```

Comment:
An error seems unlikely here, but I'm wondering if it should be checked?

Reply:
Sure, checked in 777e0ae